syncer/syncer.go

```diff
@@ -924,15 +923,15 @@ func (s *Syncer) syncDDL(tctx *tcontext.Context, queueBucket string, db *DBConn,
 					s.tctx.L().Warn("skip shard DDL handle in optimistic shard mode", zap.Strings("ddl", sqlJob.ddls))
 				}
 			} else if s.optimist.PendingOperation() == nil {
-				err = fmt.Errorf("missing shard DDL lock operation for shard DDL info (%s)", shardInfo)
+				err = terror.ErrWorkerDDLLockOpNotFound.Generate(shardInfo)
 			} else {
 				err = s.optimist.DoneOperation(*(s.optimist.PendingOperation()))
 			}
 		}
 		s.jobWg.Done()
 		if err != nil {
-			s.execErrorDetected.Set(true)
 			if !utils.IsContextCanceledError(err) {
+				s.execErrorDetected.Set(true)
```

Comment:
IMO, for `context.Canceled` error we also should set `execErrorDetected` to true so that the other functions who probe this variable can quit ASAP. If we don't set `execErrorDetected` to true syncer may still think everything works fine.